### PR TITLE
(#5442) fetch should also resolve canonical URL references

### PR DESCRIPTION
**Pull request by TahaAttari**
I tried to follow the existing pattern, I'm open to any suggestions to make this less ugly

### CODE REVIEW

File: hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	private static DefaultProfileValidationSupport myDefaultValidationSupport = new DefaultProfileValidationSupport(ourCtx);
+	private static ValidatorResourceFetcher fetcher;
+	private static DaoRegistry mockDaoRegistry;
+	private static IFhirResourceDao mockResourceDao;
+
+	@BeforeEach
+	public void before() {
+    mockDaoRegistry = mock(DaoRegistry.class);
+    mockResourceDao = mock(IFhirResourceDao.class);
+    String resource;
```
Comment (jmarchionatto) [inline hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java:41]: suggestion: use ClassPathUtil which includes not found handling

File: hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.hl7.fhir.r5.elementmodel.Element;
+import org.hl7.fhir.r5.utils.XVerExtensionManager;
+import org.hl7.fhir.validation.instance.InstanceValidator;
+import org.hl7.fhir.common.hapi.validation.validator.VersionSpecificWorkerContextWrapper;
+import ca.uhn.fhir.jpa.api.dao.IFhirResourceDao;
+
+
+public class ValidatorResourceFetcherTest extends BaseTest {
```
Comment (jmarchionatto) [inline hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java:30]: Please implement IJ suggestions to eliminate warnings

**Review by jmarchionatto — DISMISSED** [review]
Looks good. Approved with a few suggestions

**Comment by jmarchionatto** [comment]
Please also add a changelog

**Review by jmarchionatto — APPROVED** [review]

File: hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java (hunk 2 above)
Comment (TahaAttari) [reply hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java:30]: Should I implement the IJ suggestions on `ValidatorResourceFetcher` as well? There are a few redundancies

Comment (jmarchionatto) [reply hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java:30]: Please do if it is simple. Less warnings make an important warning more visible. Thanks

Comment (TahaAttari) [reply hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java:31]: done

File: hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java (hunk 1 above)
Comment (TahaAttari) [reply hapi-fhir-storage/src/test/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcherTest.java]: done

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcher.java [hunk 3]
```diff
@@ -85,6 +94,24 @@ public Element fetch(IResourceValidator iResourceValidator, Object appContext, S
...
+		CanonicalPair pair = new CanonicalPair(url);
+		SearchParameterMap searchParameterMap = new SearchParameterMap();
+		searchParameterMap.add("url", new UriParam(pair.getUrl()));
+		String version = pair.getVersion();
+		if (version != null && !version.isEmpty()) {
+			searchParameterMap.add("version", new TokenParam(version));
+		}
+		List<IBaseResource> results =
+				dao.search(searchParameterMap, requestDetails).getAllResources();
+		if (results.size() > 0) {
```
[the remark this answers is not in the document]
Comment (TahaAttari) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcher.java:100]: done

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcher.java [hunk 4]
```diff
@@ -65,18 +70,22 @@ public ValidatorResourceFetcher(
 
 	@Override
 	public Element fetch(IResourceValidator iResourceValidator, Object appContext, String theUrl)
-			throws FHIRFormatError, DefinitionException, FHIRException, IOException {
+			throws FHIRFormatError, DefinitionException, FHIRException {
 		IdType id = new IdType(theUrl);
 		String resourceType = id.getResourceType();
 		IFhirResourceDao<?> dao = myDaoRegistry.getResourceDao(resourceType);
-		IBaseResource target;
+		IBaseResource target = null;
```
[the remark this answers is not in the document]
Comment (TahaAttari) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcher.java]: done

File: hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcher.java [hunk 5]
```diff
@@ -85,15 +87,40 @@ public Element fetch(IResourceValidator iResourceValidator, Object appContext, S
...
+	private IBaseResource fetchByUrl(String url, IFhirResourceDao<?> dao, RequestDetails requestDetails)
+			throws ResourceNotFoundException {
+		CanonicalPair pair = new CanonicalPair(url);
+		SearchParameterMap searchParameterMap = new SearchParameterMap();
+		searchParameterMap.add("url", new UriParam(pair.getUrl()));
+		String version = pair.getVersion();
+		if (version != null && !version.isEmpty()) {
+			searchParameterMap.add("version", new TokenParam(version));
+		}
+		List<IBaseResource> results = null;
```
Comment (TahaAttari) [inline hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcher.java:99]: I know this is redundant but I get a linter warning if I don't do it

Comment (jmarchionatto) [reply hapi-fhir-storage/src/main/java/ca/uhn/fhir/jpa/validation/ValidatorResourceFetcher.java:99]: Really? Never mind, then.

**Pull request merged**